### PR TITLE
Fetch User Details from Discord Id

### PR DESCRIPTION
**Pull request by j24m**
## Issue Ticket : 
- https://github.com/Real-Dev-Squad/discord-slash-commands/issues/106
## PR 
- To be checked along with this feature :  
https://github.com/Real-Dev-Squad/discord-slash-commands/pull/132
## Description
- This pull request adds a new feature to the **users** module that allows users to be retrieved by their Discord ID.
- The following changes have been made : 
-> The ``getUsers( )`` middleware has been updated to support the ``discordId`` query parameter.
-> The ``retrieveUsers( )`` function checks if the discordId parameter is present in the request data.
-> The ``fetchUser( )`` service has been updated to support the ``discordId`` option.
## Tests
- Unit and integration tests have been updated to cover the new functionality.
**Unit test stats(Before) :**
image

**Unit test stats(After) :** 
image
**Integration test stats(Before) :**
image

**Integration test stats(After) :** 
image

## Working proof 
- API to fetch User details by passing discord Id as query params : localhost:3000/users/?**dev=true&discordId=504855562094247953**

https://github.com/Real-Dev-Squad/website-backend/assets/53934353/a9682cb2-e3eb-46cc-b544-4c22148ade89

### CODE REVIEW

**Review by iamitprakash — CHANGES_REQUESTED** [review]
Why coverage is so low, can you please add before code change stats as well

**Review by iamitprakash — CHANGES_REQUESTED** [review]
Why coverage is so low, can you please add before code change stats as well

**Review by iamitprakash — CHANGES_REQUESTED** [review]
where is working example proof ..?

**Review by Pratiyushkumar — APPROVED** [review]

File: controllers/users.js [hunk 1]
```diff
@@ -128,6 +128,30 @@ const getUsers = async (req, res) => {
       }
     }
 
+    // getting user details by discord id if present.
+    const dev = req.query.dev === "true";
+    if (dev) {
+      if (req.query.discordId) {
+        const discordId = req.query.discordId;
```
Comment (iamitprakash) [inline controllers/users.js:135]: in the line const discordId = req.query.discordId;. The req.query object is a JavaScript object that contains the query parameters passed in the URL. In this case, the query parameter is discordId. However, the code is trying to access the discordId property of the req.query object before it has been initialized. This will cause an error.

File: middlewares/validators/user.js [hunk 2]
```diff
@@ -158,6 +159,9 @@ async function getUsers(req, res, next) {
       id: joi.string().optional().messages({
         "string.empty": "id value must not be empty",
       }),
+      discordId: joi.string().optional().messages({
```
Comment (iamitprakash) [inline middlewares/validators/user.js:162]: messages object. The string.empty message is only triggered if the discordId value is empty. However, the optional() modifier means that the discordId value is allowed to be empty. This means that the string.empty message will never be triggered.

File: test/unit/models/users.test.js [hunk 3]
```diff
@@ -117,6 +117,24 @@ describe("users", function () {
     });
   });
 
+  describe("fetch user details based on discord id", function () {
```
Comment (iamitprakash) [inline test/unit/models/users.test.js:120]: here expect(result.user).to.haveOwnProperty("state");. The result.user object is the result of the users.fetchUser() function, which returns a user object. The state property is not part of the user object, so the expect() assertion will fail.

To fix this, you need to remove the expect(result.user).to.haveOwnProperty("state"); assertion.

File: models/users.js [hunk 4]
```diff
@@ -259,6 +259,17 @@ const fetchUser = async ({ userId = null, username = null, githubUsername = null
         id = doc.id;
         userData = doc.data();
       });
+    } else if (discordId) {
+      const user = await userModel.where("discordId", "==", discordId).limit(1).get();
+      user.forEach((doc) => {
+        id = doc.id;
+        userData = doc.data();
+      });
+      const userStatusDocs = await userStatusModel.where("userId", "==", id).limit(1).get();
```
Comment (iamitprakash) [inline models/users.js:268]: here const userStatusDocs = await userStatusModel.where("userId", "==", id).limit(1).get();. The userStatusDocs variable is a list of documents. However, the userStatusDoc variable is only initialized with the first document in the list. This means that the data variable will only contain the data from the first document.

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: controllers/users.js [hunk 5]
```diff
@@ -128,6 +128,30 @@ const getUsers = async (req, res) => {
...
+
+        let result, user;
+        try {
+          result = await dataAccess.retrieveUsers({ discordId: discordId });
```
Comment (DashDeipayan) [inline controllers/users.js:139]: can be written as `result = await dataAccess.retrieveUsers({ discordId });`

File: controllers/users.js [hunk 6]
```diff
@@ -128,6 +128,30 @@ const getUsers = async (req, res) => {
...
+          user = result.user;
+        } catch (error) {
+          logger.error(`Error while fetching user: ${error}`);
+          return res.boom.serverUnavailable(SOMETHING_WENT_WRONG);
```
Comment (DashDeipayan) [inline controllers/users.js:143]: I think an error of INTERNAL SERVER ERROR is a more meaningful error here

File: services/dataAccessLayer.js [hunk 7]
```diff
@@ -47,6 +48,9 @@ const retrieveUsers = async ({
       users.push(user);
     });
     return { users, nextId, prevId };
+  } else if (discordId !== null) {
+    result = await userQuery.fetchUser({ discordId: discordId });
```
Comment (DashDeipayan) [inline services/dataAccessLayer.js:52]: same as above

**Review by DashDeipayan — CHANGES_REQUESTED** [review]

File: test/unit/models/users.test.js (hunk 3 above)
Comment (j24m) [reply test/unit/models/users.test.js:120]: image
I am adding the state property in the beforeEach block.

Comment (j24m) [reply test/unit/models/users.test.js:120]: This is the updated user object after adding the state property in the beforeEach block. 
image

File: models/users.js (hunk 4 above)
Comment (prakashchoudhary07) [inline models/users.js:268]: Why getting userStatus details in retrieve user by discordId

**Review by iamitprakash — APPROVED** [review]

**Review by DashDeipayan — DISMISSED** [review]

**Review by iamitprakash — CHANGES_REQUESTED** [review]
please change as per Prakash change request

File: services/dataAccessLayer.js [hunk 8]
```diff
@@ -47,6 +48,9 @@ const retrieveUsers = async ({
       users.push(user);
     });
     return { users, nextId, prevId };
+  } else if (discordId !== null) {
+    result = await userQuery.fetchUser({ discordId });
+    return result;
```
Comment (prakashchoudhary07) [inline services/dataAccessLayer.js:53]: Please filter the result according to the levels here
Do not directly return the result

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

File: services/dataAccessLayer.js [hunk 9]
```diff
@@ -12,8 +13,8 @@ const retrieveUsers = async ({
   role = null,
   userIds = null,
 }) => {
+  let result = {};
```
Comment (prakashchoudhary07) [inline services/dataAccessLayer.js:16]: Why initialize result as an empty object?

**Comment by iamitprakash** [comment]
@j24m  please add working proof

**Review by gauravsinhaweb — APPROVED** [review]

**Review by gauravsinhaweb — CHANGES_REQUESTED** [review]
Add updated test result

**Comment by Achintya-Chatterjee** [comment]
Please add test stats

File: controllers/users.js [hunk 10]
```diff
@@ -130,6 +132,38 @@ const getUsers = async (req, res) => {
       }
     }
 
+    // getting user details by discord id if present.
+    const discordId = req.query.discordId;
+
+    const dev = req.query.dev === "true";
+    if (req.query.discordId) {
```
Comment (iamitprakash) [inline controllers/users.js:139]: can we merge both condition #139 and #140

File: controllers/users.js [hunk 11]
```diff
@@ -130,6 +132,38 @@ const getUsers = async (req, res) => {
...
+      if (dev) {
```
Comment (iamitprakash) [inline controllers/users.js:140]: dev seems to sting into line 138, then why it's like bool

File: controllers/users.js [hunk 12]
```diff
@@ -130,6 +132,38 @@ const getUsers = async (req, res) => {
...
+        let result, user;
+        try {
+          result = await dataAccess.retrieveUsers({ discordId });
+          user = result.user;
```
Comment (iamitprakash) [inline controllers/users.js:144]: what if result is null or empty

Comment (iamitprakash) [inline controllers/users.js:144]: please add test case to see user array is null

File: controllers/users.js (hunk 10 above)
Comment (j24m) [reply controllers/users.js:139]: done

File: controllers/users.js (hunk 12 above)
Comment (j24m) [reply controllers/users.js:144]: done

**Review by gauravsinhaweb — APPROVED** [review]

**Comment by iamitprakash** [comment]
@j24m  
image

**Review by sahsisunny — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]
